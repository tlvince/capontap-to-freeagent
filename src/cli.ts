#!/usr/bin/env node

import { convert } from '.'

let input = ''
process.stdin.on('data', data => {
  input += data
})

process.stdin.once('end', () => {
  convert(input)
    .then(res => process.stdout.write(res))
    .catch(err => process.stderr.write(err))
})
