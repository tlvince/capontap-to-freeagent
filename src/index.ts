import neatCsv, { Row } from 'neat-csv'

const freeAgentHeaders = ['Date', 'Amount', 'Description']

const isValid = (row: Row): boolean =>
  freeAgentHeaders.every(
    (freeAgentHeader: string): string => row[freeAgentHeader]
  )

const inverseAmount = (str: string) => {
  if (str === '0') return str
  if (str.startsWith('-')) return str.slice(1)
  return '-' + str
}

const transform = (row: Row) => ({
  ...row,
  Amount: inverseAmount(row.Amount),
})

export const convert = async (csvLines: string): Promise<string> => {
  const rows = await neatCsv(csvLines)
  return rows
    .filter(isValid)
    .filter(Boolean)
    .map(transform)
    .reduce((acc, row) => {
      return acc + freeAgentHeaders.map(header => row[header]).join(',') + '\n'
    }, '')
}
