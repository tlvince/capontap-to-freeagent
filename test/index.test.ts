import { convert } from '../src'
import { promises as fsP } from 'fs'

interface SetupResponse {
  csvFixture: string
}

const setup = async (): Promise<SetupResponse> => {
  const buffer = await fsP.readFile('./test/fixtures/capontap.csv')
  const csvFixture = buffer.toString()
  return {
    csvFixture,
  }
}

describe('convert', () => {
  it('should handle a well-formed CSV', async () => {
    const { csvFixture } = await setup()
    const converted = await convert(csvFixture)
    expect(converted).toMatchSnapshot()
  })
})
