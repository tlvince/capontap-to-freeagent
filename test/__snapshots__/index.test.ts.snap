// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should handle a well-formed CSV 1`] = `
"20/10/2019,-123.45,Waitrose
18/10/2019,2.20,Amazon refund
27/09/2019,0,Finance charge
"
`;
